csharp/ql/lib/semmle/code/csharp/dataflow/internal/ContentDataFlow.qll

```diff
@@ -1,185 +1,215 @@
-   * class MyAnalysisConfiguration extends ContentDataFlowConfiguration {
-   *   MyAnalysisConfiguration() { this = "MyAnalysisConfiguration" }
-   *   // Override `isSource` and `isSink`.
-   *   // Optionally override `isBarrier`.
-   *   // Optionally override `isAdditionalFlowStep`.
-   *   // Optionally override `getAFeature`.
-   *   // Optionally override `accessPathLimit`.
-   *   // Optionally override `isRelevantContent`.
-   * }
-   * ```
-   *
-   * Unlike `DataFlow::Configuration` (on which this class is based), we allow
-   * for data to be stored (possibly nested) inside contents of sources and sinks.
-   * We track flow paths of the form
-   *
-   * ```
-   * source --value-->* node
-   *        (--read--> node --value-->* node)*
-   *        --(non-value|value)-->* node
-   *        (--store--> node --value-->* node)*
-   *        --value-->* sink
-   * ```
-   *
-   * where `--value-->` is a value-preserving flow step, `--read-->` is a read
-   * step, `--store-->` is a store step, and `--(non-value)-->` is a
-   * non-value-preserving flow step.
+   * Gets a data flow configuration feature to add restrictions to the set of
+   * valid flow paths.
    *
-   * That is, first a sequence of 0 or more reads, followed by 0 or more additional
-   * steps, followed by 0 or more stores, with value-preserving steps allowed in
-   * between all other steps.
+   * - `FeatureHasSourceCallContext`:
+   *    Assume that sources have some existing call context to disallow
+   *    conflicting return-flow directly following the source.
+   * - `FeatureHasSinkCallContext`:
+   *    Assume that sinks have some existing call context to disallow
+   *    conflicting argument-to-parameter flow directly preceding the sink.
+   * - `FeatureEqualSourceSinkCallContext`:
+   *    Implies both of the above and additionally ensures that the entire flow
+   *    path preserves the call context.
    */
-  abstract class Configuration extends string {
-    bindingset[this]
-    Configuration() { any() }
+  default DataFlow::FlowFeature getAFeature() { none() }
 
-    /**
-     * Holds if `source` is a relevant data flow source.
-     */
-    abstract predicate isSource(Node source);
+  /** Gets a limit on the number of reads out of sources and number of stores into sinks. */
+  default int accessPathLimit() { result = DataFlowPrivate::accessPathLimit() }
 
-    /**
-     * Holds if `sink` is a relevant data flow sink.
-     */
-    abstract predicate isSink(Node sink);
+  /** Holds if `c` is relevant for reads out of sources or stores into sinks. */
+  default predicate isRelevantContent(DataFlow::ContentSet c) { any() }
+}
 
-    /**
-     * Holds if data may flow from `node1` to `node2` in addition to the normal data-flow steps.
-     */
-    predicate isAdditionalFlowStep(Node node1, Node node2) { none() }
+/**
+ * Constructs a global content data flow computation.
+ */
+module Global<ContentConfigSig ContentConfig> implements DataFlow::GlobalFlowSig {
```

Comment:
Is the `implements` bit needed (or desired)?

Reply:
No, not if we only want to expose the specialized `flow` predicate.